### ISSUE TITLE
FR: Avoid enabling IP forwarding on the host

### ISSUE THREAD

**Issue by ccdunder** [Maroka-chan/VPN-Confinement#18]
Enabling IP forwarding increases the attack service of any machine. I don't believe it's necessary.

@Maroka-chan do you have any reservations about merging such a change?

**Comment by Maroka-chan** [comment]
I've also had concerns about it, but IP forwarding seems to be necessary to pre-route packets to the networking namespace :(
I haven't found a way to only allow forwarding to the netns.. It'd be interesting to see what Docker does.
I'd definitely be open to a solution that doesn't require IP forwarding :)

**Comment by unhitched9271** [comment]
just thinking out loud.. i see a few possibilities - could just enable it only on specific interfaces, rather than globally.. eg:

 ```nix
   boot.kernel.sysctl = {
     "net.ipv4.ip_forward" = 0;
     "net.ipv4.conf.all.forwarding" = 0;
     "net.ipv4.conf.default.forwarding" = 0;

     "net.ipv4.conf.**wg-br**.forwarding" = 1;
     "net.ipv4.conf.**veth-wg-br**.forwarding" = 1;
   };
 ```

if you wanted to get rid of forwarding entirely, you could copy the docker approach - only allow traffic to/from the namespace bridge, and block everything else... or.... could use a userspace service to forward the traffic back and forth (I think?) eg socat / haproxy, listening on host and forwarding to the namespace?

**Comment by unhitched9271** [comment]
I had another idea on this, which I'm trying out in a fork - instead of the approaches above, I discovered that systemd-socket-proxyd is a thing... which can forward connections from host to the vpn namespace - which perfect for accessing webui's *and* given its part of systemd.. no additional dependencies!

so the benefits of this approach are:

- no global IP forwarding required - can just enable them for the vpn interface
- still allows minimal port mappings between the namespace / host for the webui via tcp/udp
- systemd-socket-proxyd  allows socket activation of the proxy - so the proxy only runs on demand, ie when there is traffic flowing between the two

once I've spent some more time testing it, is there any interest in me uploading the patch contents @Maroka-chan ?

**Comment by Maroka-chan** [comment]
I'd love to see what you come up with!

I looked into how docker approaches this at some point, and believe I concluded that they use a proxy, which can be disabled, but in that case they also just use ip forwarding. It's been a while since I looked into it, so don't quite me on that.
Using a proxy comes with a performance penalty, so it might be nice to do both interface specific forwarding *and* systemd-socket-proxyd? Then users can choose which to use.

What are your thoughts? I doubt performance will be a problem for most usecases, so using the proxy could be the default.

**Comment by unhitched9271** [comment]
yeah anyrthing thats kernel to userspace/vice-versa will have a performance impact, how much of one i'm not sure - just as once I realised you could enable forwarding per interface rather than globally (so just for the vpn) it didn't seem that important for a webui access method. 
not had a chance to test it anymore yet, but will share what I've come up with once I'm certain i've not broken anything!
